File: schema/go/easemlschema/schema/schema_test.go
package schema

import (
	"encoding/json"
	"io/ioutil"
	"log"
	"path"
	"strings"
	"testing"
)

// Here we assume all test will be ran from the v0.1/go directory.
const relTestPath = "../../../test-examples"
const schemaValPositivePath = "schema/validate/positive"
const schemaValNegativePath = "schema/validate/negative"
const schemaMatchPositivePath = "schema/match/positive"
const schemaMatchNegativePath = "schema/match/negative"

func TestSchemaValidate(t *testing.T) {

	// Load positive test examples.
	files, err := ioutil.ReadDir(path.Join(relTestPath, schemaValPositivePath))
	if err != nil {
		log.Fatal(err)
	}
	for i := range files {
		var testName = path.Join(schemaValPositivePath, files[i].Name())
		var exampleFile = path.Join(relTestPath, schemaValPositivePath, files[i].Name())
		t.Run(testName, testSchemaValidate(exampleFile, true))
	}

	// Load negative test examples.
	files, err = ioutil.ReadDir(path.Join(relTestPath, schemaValNegativePath))
	if err != nil {
		log.Fatal(err)
	}
	for i := range files {
		var testName = path.Join(schemaValNegativePath, files[i].Name())
		var exampleFile = path.Join(relTestPath, schemaValNegativePath, files[i].Name())
		t.Run(testName, testSchemaValidate(exampleFile, false))
	}

}

func testSchemaValidate(exampleFile string, positiveExample bool) func(*testing.T) {
	return func(t *testing.T) {
		// Load the file.
		srcData, err := ioutil.ReadFile(exampleFile)
		if err != nil {
			panic(err)
		}

		// Unmarshal JSON into a map.
		var src map[string]interface{}
		err = json.Unmarshal(srcData, &src)
		if err != nil {
			panic(err)
		}

		// Load the schema.
		_, err = Load(src)
		if positiveExample && err != nil {
			schErr, ok := err.(Error)
			if ok == false {
				t.Error("Resulting error should be a schema.Error instance.")
			}
			msg := schErr.Error()
			pth := schErr.Path()

			t.Errorf("Found error '%s' on path '%s'.", msg, pth)

		} else if !positiveExample {

			if err == nil {
				t.Error("Expected an error but got none.")
			} else {
				_, ok := err.(Error)
				if ok == false {
					t.Error("Resulting error should be a schema.Error instance.")
				}
			}
		}
	}
}

func TestSchemaMatch(t *testing.T) {

	// Load positive test examples.
	files, err := ioutil.ReadDir(path.Join(relTestPath, schemaMatchPositivePath))
	if err != nil {
		log.Fatal(err)
	}
	for i := range files {
		var fileName = files[i].Name()
		if strings.HasSuffix(fileName, "-src.json") {

			fileName = strings.TrimSuffix(fileName, "-src.json")
			var fileNameSrc = fileName + "-src.json"
			var fileNameDst = fileName + "-dst.json"

			var testName = path.Join(schemaMatchPositivePath, fileName)
			var exampleFileSrc = path.Join(relTestPath, schemaMatchPositivePath, fileNameSrc)
			var exampleFileDst = path.Join(relTestPath, schemaMatchPositivePath, fileNameDst)
			t.Run(testName, testSchemaMatch(exampleFileSrc, exampleFileDst, true))
		}
	}

	// Load negative test examples.
	files, err = ioutil.ReadDir(path.Join(relTestPath, schemaMatchNegativePath))
	if err != nil {
		log.Fatal(err)
	}
	for i := range files {
		var fileName = files[i].Name()
		if strings.HasSuffix(fileName, "-src.json") {

			fileName = strings.TrimSuffix(fileName, "-src.json")
			var fileNameSrc = fileName + "-src.json"
			var fileNameDst = fileName + "-dst.json"

			var testName = path.Join(schemaMatchNegativePath, fileName)
			var exampleFileSrc = path.Join(relTestPath, schemaMatchNegativePath, fileNameSrc)
			var exampleFileDst = path.Join(relTestPath, schemaMatchNegativePath, fileNameDst)
			t.Run(testName, testSchemaMatch(exampleFileSrc, exampleFileDst, false))
		}
	}

}

func testSchemaMatch(exampleFileSrc, exampleFileDst string, positiveExample bool) func(*testing.T) {
	return func(t *testing.T) {
		// Load the file.
		srcData, err := ioutil.ReadFile(exampleFileSrc)
		if err != nil {
			panic(err)
		}
		dstData, err := ioutil.ReadFile(exampleFileDst)
		if err != nil {
			panic(err)
		}

		// Unmarshal JSON into a map.
		var src map[string]interface{}
		err = json.Unmarshal(srcData, &src)
		if err != nil {
			panic(err)
		}
		var dst map[string]interface{}
		err = json.Unmarshal(dstData, &dst)
		if err != nil {
			panic(err)
		}

		// Load the schema.
		srcSchema, err := Load(src)
		if err != nil {
			panic(err)
		}
		dstSchema, err := Load(dst)
		if err != nil {
			panic(err)
		}
		match, schMatching := dstSchema.Match(srcSchema, true)

		if positiveExample {
			if match == false {
				t.Error("Schema match failed.")
			}

			if schMatching == nil {
				t.Error("Schema matching not returned.")
			}

		} else if !positiveExample && (match == true || schMatching != nil) {

			t.Error("Expected a failure but got a match.")

		}
	}
}

File: schema/javascript/easemlschema/test/node/schema_test.js
'use strict'

import sch from '../../src/schema'
import { expect } from 'chai'

import fs from 'fs'
import path from 'path'

const relTestPath = path.join(__dirname, '../../../../test-examples')
const schemaValPositivePath = 'schema/validate/positive'
const schemaValNegativePath = 'schema/validate/negative'
const schemaMatchPositivePath = 'schema/match/positive'
const schemaMatchNegativePath = 'schema/match/negative'

describe('Schema', function () {
  describe('#validate', function () {
    describe('#positive', function () {
      var files = fs.readdirSync(path.join(relTestPath, schemaValPositivePath))
      for (let i = 0; i < files.length; i++) {
        if (files[i].endsWith('.json')) {
          let testName = path.join(schemaValPositivePath, files[i].slice(0, -'.json'.length))
          let exampleFile = path.join(relTestPath, schemaValPositivePath, files[i])
          it(testName, testSchemaValidate(exampleFile, true))
        }
      }
    })
    describe('#negative', function () {
      var files = fs.readdirSync(path.join(relTestPath, schemaValNegativePath))
      for (let i = 0; i < files.length; i++) {
        if (files[i].endsWith('.json')) {
          let testName = path.join(schemaValNegativePath, files[i].slice(0, -'.json'.length))
          let exampleFile = path.join(relTestPath, schemaValNegativePath, files[i])
          it(testName, testSchemaValidate(exampleFile, false))
        }
      }
    })
  })
  describe('#match', function () {
    describe('#positive', function () {
      var files = fs.readdirSync(path.join(relTestPath, schemaMatchPositivePath))
      for (let i = 0; i < files.length; i++) {
        if (files[i].endsWith('-src.json')) {
          let fileName = files[i].slice(0, -'-src.json'.length)
          let testName = path.join(schemaMatchPositivePath, fileName)
          let exampleFileSrc = path.join(relTestPath, schemaMatchPositivePath, fileName + '-src.json')
          let exampleFileDst = path.join(relTestPath, schemaMatchPositivePath, fileName + '-dst.json')
          it(testName, testSchemaMatch(exampleFileSrc, exampleFileDst, true))
        }
      }
    })
    describe('#negative', function () {
      var files = fs.readdirSync(path.join(relTestPath, schemaMatchNegativePath))
      for (let i = 0; i < files.length; i++) {
        if (files[i].endsWith('-src.json')) {
          let fileName = files[i].slice(0, -'-src.json'.length)
          let testName = path.join(schemaMatchNegativePath, fileName)
          let exampleFileSrc = path.join(relTestPath, schemaMatchNegativePath, fileName + '-src.json')
          let exampleFileDst = path.join(relTestPath, schemaMatchNegativePath, fileName + '-dst.json')
          it(testName, testSchemaMatch(exampleFileSrc, exampleFileDst, false))
        }
      }
    })
  })
})

function testSchemaValidate (exampleFile, positiveExample) {
  return function () {
    // Load the JSON file.
    let srcData = fs.readFileSync(exampleFile)
    let src = JSON.parse(srcData)

    if (positiveExample) {
      // Load the schema.
      let schema = sch.load(src)
      expect(schema).to.be.instanceof(sch.Schema)
    } else {
      let caughtError = null
      try {
        sch.load(src)
      } catch (error) {
        caughtError = error
        if ((caughtError instanceof sch.SchemaException) === false) {
          throw error
        }
      }
      expect(caughtError).to.be.instanceOf(sch.SchemaException)
    }
  }
}

function testSchemaMatch (exampleFileSrc, exampleFileDst, positiveExample) {
  return function () {
    // Load the JSON files.
    let srcData = fs.readFileSync(exampleFileSrc)
    let src = JSON.parse(srcData)
    let dstData = fs.readFileSync(exampleFileDst)
    let dst = JSON.parse(dstData)

    // Load the schemas.
    let srcSchema = sch.load(src)
    let dstSchema = sch.load(dst)

    if (positiveExample) {
      // Match the schemas.
      let match = dstSchema.match(srcSchema, true)
      expect(match).to.be.not.null
      expect(match).to.be.instanceof(sch.Schema)
    } else {
      // Match the schemas.
      let match = dstSchema.match(srcSchema, false)
      expect(match).to.be.a('boolean')
      expect(match).to.be.false
    }
  }
}
